### ISSUE TITLE
Support local variables (tb.tb_frame.f_locals)

### ISSUE THREAD

**Issue by iankharris** [ionelmc/python-tblib#41]
I'm sure there's a reason for this, but it seems that Traceback objects don't copy the `f_locals` dictionary from the traceback frame.  It would be nice to have this so that, for instance, reraised exceptions can be better introspected in an IDE.

For instance:
```
>>> def func():
...     a = 1
...    print locals()
...     try:
...         raise ValueError()
...     except Exception as e:
...         exc_type, exc_value, exc_tb = sys.exc_info()
>>> exc_tb.tb_frame.f_locals
{'a': 1, e: ValueError()}
>>> tb = Traceback(exc_tb)
>>> tb.as_traceback().tb_frame.f_locals
{}
```

There might be something obvious here that I'm missing?

**Comment by ionelmc** [comment]
`f_locals` are a can of worms basically, we'd have lots of corner cases like reference loops, unpickleable stuff, overblown pickle result size and such. I'm not saying no but I'm not a fan of the idea either - the primary goal of this library is to be able to display a traceback using stdlib modules (like `traceback`) from a different process.

**Comment by iankharris** [comment]
I understand.  An easy solution might be to just assign all unpickleable or otherwise problematic `f_locals` to `None` and be done with it.  I'm not sure how we can detect reference loops but I haven't done much work on code like this.

**Comment by ionelmc** [comment]
Btw, previous code and discussion on exactly what you asked: https://github.com/ionelmc/python-tblib/pull/6 - I guess you could try it and report if it works in your IDE?

**Comment by imachug** [comment]
@ionelmc Do you think we could do the following: we could only pickle simple objects (say, primitives and `list`/`dict`) to avoid space issues.

The problem is: gevent uses tblib to save exceptions when they are thrown in another thread to get rid of locals, but pytest uses f_locals for traceback. This means that gevent and pytest don't work together. Pickling primitives shouldn't increase refcount or affect performance badly but will make debugging waaay easier.

**Comment by davegravy** [comment]
Am trying to use tblib to pickle tracebacks in an application with spotty internet so I can later send to sentry via sentry_sdk.capture_exception() in a retry process. Currently all the local variables are blank and this feature would be really handy even if it only supported primitives. 

Is there anything other than docs and tests blocking [#59](https://github.com/ionelmc/python-tblib/pull/59)?

**Comment by ionelmc** [comment]
So what I would like to have here is some sort of parametrization in tblib.pickle_support's pickle_traceback and install. I think that ideally install would take an option/hook to customize the locals attribute (so the user decides what ends up in locals via his hook). Then I won't need to have a hack in tblib for some pytest thingy or whatever.

Eg:

```python
from tblib import pickling_support

def pytest_locals(frame):
    if "__tracebackhide__" in frame.f_locals:
        return {"__tracebackhide__": bool(frame.f_locals["__tracebackhide__"])}
    else:
        return {}

pickling_support.install(pickle_locals=pytest_locals)
```

**Comment by hoodmane** [comment]
Yes, this proposal sounds great.

**Comment by ionelmc** [comment]
For your use-case it's going to be roughly like:

```py
from tblib import pickling_support

def get_pytest_locals(frame):
    if frame.f_locals.get("__tracebackhide__"):
        return {"__tracebackhide__": True}
    else:
        return {}

pickling_support.install(get_locals=get_pytest_locals)
```

**Comment by hoodmane** [comment]
Thanks!

**Comment by ionelmc** [comment]
Released 3.0
